packages/number-field/src/NumberField.ts

```diff
@@ -495,7 +495,17 @@ export class NumberField extends TextfieldBase {
         // Step shouldn't validate when 0...
         if (this.step) {
             const min = typeof this.min !== 'undefined' ? this.min : 0;
-            const moduloStep = (value - min) % this.step;
+            const digitsAfterDecimals =
+                this.step != Math.floor(this.step)
+                    ? this.step.toString().split('.')[1].length
+                    : 0;
+            const formatter = new Intl.NumberFormat('en-US', {
+                maximumFractionDigits: digitsAfterDecimals,
+            });
```

Comment:
This news a NumberFormat for _every_ validation. It really only needs to do that for each time the value of `step` changes. In the case that we keep a _second_ formatter like this, we likely want to cache it's value in some way. Similarly, we could add this logic to the larger construction in `protected get numberFormatter()` and only remove it's cache when updating `step` which would allow us to manage a single formatter.

Reply:
Another approach that comes to my mind is to cache the formatter and only create a new one when `digitsAfterDecimals` changes, I can leverage a private field to store the formatter and another one to store the last known value of `digitsAfterDecimals`. Then, in the getter for the formatter, I can check if `digitsAfterDecimals` has changed and create a new formatter if it has changed.

`private _formatter: Intl.NumberFormat | null = null;
private _lastDigitsAfterDecimals: number | null = null;

protected get formatter(): Intl.NumberFormat {
    if (this._formatter === null || this._lastDigitsAfterDecimals !== this.digitsAfterDecimals) {
        this._formatter = new Intl.NumberFormat('en-US', {
            maximumFractionDigits: this.digitsAfterDecimals,
        });
        this._lastDigitsAfterDecimals = this.digitsAfterDecimals;
    }
    return this._formatter;
}`
THOUGHTS?

Reply:
I'm fine with having two formatters if you see that as the right approach here, as long as both are cached appropriately. We should ensure that they both use the `NumberFormatter` helper rather than `Intl.NumberFormat` directly so that there isn't confusion down the line.

Reply:
I had to use 2 formatters because when I was combining the logic in a single formatter, there were a lot of vrt's failing when the number field had units like percentage or dollar etc, because that formatter was called twice once inside `validateInput` and once to maintain the units.